src/opnsense/mvc/app/models/OPNsense/Base/Constraints/ComparedToFieldConstraint.php

```diff
@@ -54,23 +54,23 @@ public function validate($validator, $attribute): bool
         $node = $this->getOption('node');
         $field_name = $this->getOption('field');
         $operator = $this->getOption('operator');
-        if ($node && !($this->isEmpty($node) || empty($operator) || empty($field_name))) {
+        if ($node && !($node->isEmpty() || empty($operator) || empty($field_name))) {
             $parent_node = $node->getParentNode();
             $other_node_content = $parent_node->$field_name;
 
             // if the other field is not set, or invalid type -> ignore this constraint
             if (
-                $this->isEmpty($other_node_content) ||
-                    !is_numeric((string)$node) && !is_numeric((string)$other_node_content)
+                is_null($other_node_content) || $other_node_content->isEmpty() ||
+                !$node->isNumeric() && !$other_node_content->isNumeric()
```

Comment:
hum, probably `!$other_node_content->isNumeric()` may have issue when `$other_node_content` is `null`. Wondering what would happen with the `&&` in the middle here

Reply:
It’s a bit of a trick question. Null is not numeric, but it’s certainly empty. But ‘0’ is numeric, but still empty. The practicality of the checks either needs to be guarded or we need earlier checks I discard handling invalid values or just assume null is either ‘’ or 0

Reply:
The more I read it the more I'm confused :exploding_head:  . I'm now thinking I spoke too fast :expressionless: .

If you think it should be changed, would you be kind to provide a bit of code for the required change :pray:

Reply:
I'm not sure. Just throwing ideas out at the moment. :) Could we maybe start with merging a new test case as a separate PR first so we can add the visibility for it being broken?

Reply:
> new test case

Do you have Unit Tests :heart: ? Where?

Reply:
Hmm, so in general:

    # pkg install ps-debug && cd /usr/core && make test

The actual code for the FieldTypes and Constraints in particular:

https://github.com/opnsense/core/tree/master/src/opnsense/mvc/tests/app/models/OPNsense/Base

Reply:
sorry, small typo in previous, should have been "os-debug"

Reply:
Really missed that folder thanks.

`pkg install os-debug` I'm on Linux, will try to get tests working in some way. Any specific doc I should read, https://github.com/opnsense/core/blob/master/CONTRIBUTING.md doesn't talk about that :shrug:

Reply:
`phpunit --configuration PHPunit.xml` should do the trick on other platforms as well, you can find it in `src/opnsense/mvc/tests/PHPunit.xml`.

Out of sheer curiosity, what are you building?

Reply:
Should I upgrade my opnsense VM to some `main` branch equivalent? how? Sorry for my silly questions :( 
I'm on `OPNsense 24.7.4_1-amd64`

![image](https://github.com/user-attachments/assets/fd179cdf-4438-445d-983f-8bca08300bdf)

Reply:
The tests expect some persistent installed state from the core package so it's only possible to run it in an opnsense VM at the moment. In theory you only need PHPUnit but since our tests run during a nightly build that is not a pressing thing to do (also because PHP versions may differ in results). 

Just for reference these are some debug tools the plugin pulls in https://github.com/opnsense/plugins/blob/b8699cde8cf210e864e4bfab7329dcaaf40ec995/devel/debug/Makefile#L4-L8

Reply:
@AdSchellevis 
> phpunit --configuration PHPunit.xml

That made it!
![image](https://github.com/user-attachments/assets/21814f2e-943e-46f2-983c-ec82b80224d8)


> Out of sheer curiosity, what are you building?

I first manually installed [`restic`](https://restic.readthedocs.io/en/stable/030_preparing_a_new_repo.html) to backup my configuration, but found i need to create some action to make it scheduled and needed some script and password... This wasn't much pratical to restore in case of disaster recovery and having a plugin would be better. plus it would certainly help others

I'm currently having pages and configuration saving working
![image](https://github.com/user-attachments/assets/fa3cf75e-9e52-40b3-8923-e160adba571b)
![image](https://github.com/user-attachments/assets/e3630831-80ec-4918-bc29-5eff11114434)
![image](https://github.com/user-attachments/assets/04bb4498-5fbe-42a1-9190-215cacc5a1e8)
also templates and a basic script
![image](https://github.com/user-attachments/assets/4cb7bd9c-8128-43e9-bcb6-0c54a708efae)

I will next work on  creating cron/schedules like in acme plugin. And also wish to integrate it in some way to trigger an automatic backup after each config change.

Reply:
Please find PR #7883 for the preliminary Test Case

Reply:
So I still think `!$other_node_content->isNumeric()` may overlap with `$other_node_content->isEmpty()` in the "0" case. I'm not sure what the best way to untangle this is. But I admit I also don't know if that is actually an issue. @AdSchellevis what do you think?

Reply:
@fichtner I think the case you are thinking of is `node=2`, `othernode=0` `operator=lt`, and you are expecting an error. (Current code in `master` don't raise any error message for that - neither "next" one)

Test would be:
```php
    public function test_if_it_validates_zero_number_ranges_correctly_with_lt_and_error()
    {
        $validator = new \OPNsense\Base\Validation();
        $validate = $this->make_validator(2, 0, 'test', 'lt');
        $ret = $validate->validate($validator, '');
        $messages = $validator->getMessages();
        $this->assertEquals(1, count($messages));
        $this->assertEquals(true, $ret);
    }
```

Reply:
I pushed new commit https://github.com/opnsense/core/pull/7879/commits/1651ff82f10cc548e1a19f8efdd3dc5288fbad54 to fix that specific case. It introduces a new `isEmptyButZero()` function. Let me know what you think.